### ISSUE TITLE
runtime/pprof: TestCPUProfileMultithreadMagnitude flaky due to Linux kernel bug

### ISSUE THREAD

**Issue by prattmic** [golang/go#49065]
[2021-10-19T07:45:46-ee92daa/linux-386-longtest](https://build.golang.org/log/96de4e4aa99b8f02198122daf296bfcbbdb76fd7)
[2021-10-18T21:21:54-6c0daa7/linux-386-longtest](https://build.golang.org/log/cd243d02a622b97f575d8f938e3029c45419790a)
[2021-10-18T19:46:27-394a1ad/linux-amd64-longtest](https://build.golang.org/log/98427a8ba9869e3d2dea30e65757798c0eb00791)
[2021-10-15T20:29:48-85cbdda/linux-riscv64-jsing](https://build.golang.org/log/f50dfeb1f5fa762ad5f58fe2a288d2f00fba8a5f)
[2021-10-12T20:20:41-d032b2b/linux-amd64-wsl](https://build.golang.org/log/4ad25d9ce2cc4a182951c0a7a4a99038dd68aa25)
[2021-10-11T12:53:11-34f7b1f/linux-riscv64-unmatched](https://build.golang.org/log/febc7cc0aa74c4f3a439f4ea8b964cb35b6b3c5c)
[2021-10-08T17:32:31-d7ba1d2/linux-riscv64-unmatched](https://build.golang.org/log/ec7481aad121a0a9624a9f14703be85a17c737fc)
[2021-10-08T14:08:12-5b9206f/linux-riscv64-unmatched](https://build.golang.org/log/8b21be3e09bd55ec3c56cdae1c17a5ae98a0ce0f)
[2021-10-06T16:11:20-aecf4b1/linux-riscv64-unmatched](https://build.golang.org/log/fd37315fcd67a868b21b4c3e8f4568f6fafc0fb8)
[2021-10-05T16:06:49-060cd73/linux-riscv64-unmatched](https://build.golang.org/log/27402cdd15681ddd46e7f4072e2457254c45d9af)
[2021-10-02T13:52:59-07a7c6f/linux-riscv64-unmatched](https://build.golang.org/log/ce9dc2fc76c8823b54b226451588e32411fcf7b8)
[2021-10-01T17:27:29-3357624/linux-riscv64-unmatched](https://build.golang.org/log/0571ceabcdf23d3d7ff0b25272664ba7d95751df)
[2021-10-01T17:27:24-952df98/linux-riscv64-jsing](https://build.golang.org/log/8567acd585974b3a978af5f45184faa3c4a09491)
[2021-10-01T04:30:22-822f349/linux-amd64-wsl](https://build.golang.org/log/4df94193694d5e5b62e14637df3f0ad2c04fce68)
[2021-09-30T16:43:32-9112d29/linux-riscv64-unmatched](https://build.golang.org/log/dda19bb3d2ab1bb6967fc9f5ce7ccbdaa6b18f2a)
[2021-09-27T18:58:36-5b90958/linux-amd64-longtest](https://build.golang.org/log/a712ddb9bd0ebd3345549fb5bf67118fce6550b8)
[2021-09-27T18:58:36-5b90958/linux-amd64-wsl](https://build.golang.org/log/49a9d72637812feab375613b6746d608f7cabe74)
[2021-09-27T18:58:29-8cfd8c3/linux-amd64-longtest](https://build.golang.org/log/f1b9a22b2fef09e5219f117ee3db86154064e99e)
[2021-09-27T18:58:29-8cfd8c3/linux-amd64-wsl](https://build.golang.org/log/d5cf3e09d1e320b448ccf26af5d7bb05c561247d)

Note that many of these include other profiling test failures which may indicate an issue with profiling beyond a fragile test.

cc @rhysh

**Comment by rhysh** [comment]
Thank you for the links. Here's what I've found so far in the 386 and amd64 (not WSL, not RISC-V) logs.

The failing test results for TestCPUProfileMultithreadMagnitude on those platforms all look like "got full profiling info, but only from a subset of threads", followed by some number of tests that get a very small number of profile samples—often 0, sometimes 1, sometimes dozens—when the tests expect several hundred. The samples that do appear come from the runtime, in the scheduler or scavenger, in functions that might run on otherwise-idle threads.

---

386

---

[2021-10-19T07:45:46-ee92daa/linux-386-longtest](https://build.golang.org/log/96de4e4aa99b8f02198122daf296bfcbbdb76fd7)

HOSTNAME=buildlet-linux-stretch-morecpu-rn873c60c, which per https://farmer.golang.org/builders is "Debian Stretch, but on e2-highcpu-16", which per https://cloud.google.com/compute/docs/general-purpose-machines#e2-high-cpu is 16 vCPUs and 16 GB memory.

```
    pprof_test.go:541: runtime/pprof.cpuHog1: 498
    pprof_test.go:541: runtime/pprof.cpuHog3: 3471
    --- FAIL: TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof (0.00s)
        pprof_test.go:179: compare 1m19.558967s vs 34.71s
        pprof_test.go:191: CPU usage reports are too different (limit -10.0%, got -56.4%)
```

The single-goroutine portion of TestCPUProfileMultithreadMagnitude saw 4.98 seconds of samples (in a 5-second test). The parallel portion (using GOMAXPROCS goroutines) also ran for 5 seconds, during which the OS reported 79.56 seconds of user+system time in rusage but the runtime only saw 34.71 seconds worth of samples. That looks like "the cliff at 2.5 that we got with setitimer is now present at 7 vCPU, sometimes".

The failure in that test run of TestCPUProfileMultithreadMagnitude is followed by failures in TestCPUProfileInlining, TestCPUProfileRecursion, TestMathBigDivide, TestMorestack, TestCPUProfileLabel, and TestTimeVDSO. All of those have "total 0 CPU profile samples collected" except for TestCPUProfileInlining which got a single sample (below the minimum of 125 and target of 500).

---

[2021-10-18T21:21:54-6c0daa7/linux-386-longtest](https://build.golang.org/log/cd243d02a622b97f575d8f938e3029c45419790a)

HOSTNAME=buildlet-linux-stretch-morecpu-rnce27d47, so GOMAXPROCS=16 again.

```
    pprof_test.go:541: runtime/pprof.cpuHog1: 499
    pprof_test.go:541: runtime/pprof.cpuHog3: 2994
    --- FAIL: TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof (0.00s)
        pprof_test.go:179: compare 1m19.983449s vs 29.94s
        pprof_test.go:191: CPU usage reports are too different (limit -10.0%, got -62.6%)
```

The single-goroutine portion saw 4.99 seconds of samples, and the parallel portion saw 29.94 seconds of samples. This looks like "the cliff at 2.5 that we got with setitimer is now present at 6 vCPU, sometimes".

The failure in that test run of TestCPUProfileMultithreadMagnitude is followed by failures in TestCPUProfileRecursion, TestMathBigDivide, TestMorestack, TestCPUProfileLabel, and TestTimeVDSO. Again all of those failures were with "total 0 CPU profile samples collected" except for TestCPUProfileRecursion and TestMorestack which got a single sample each (vs min of 125 and target of 500).

---

amd64

---

[2021-10-18T19:46:27-394a1ad/linux-amd64-longtest](https://build.golang.org/log/98427a8ba9869e3d2dea30e65757798c0eb00791)

No environment header at the top of this log.

```
    pprof_test.go:541: runtime/pprof.cpuHog1: 500
    pprof_test.go:541: runtime/pprof.cpuHog3: 4987
    --- FAIL: TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof (0.00s)
        pprof_test.go:179: compare 1m19.876446s vs 49.87s
        pprof_test.go:191: CPU usage reports are too different (limit -10.0%, got -37.6%)
```

The single-goroutine portion saw 5.00 seconds of samples, and the parallel portion saw 49.87 seconds of samples. This looks like "the cliff at 2.5 that we got with setitimer is now present at 10 vCPU, sometimes".

Followed by failure of TestMorestack (10 samples, 0 where expected), TestCPUProfileLabel (22 samples, 0 where expected), and TestTimeVDSO (9 samples, 0 where expected). 

---

[2021-09-27T18:58:36-5b90958/linux-amd64-longtest](https://build.golang.org/log/a712ddb9bd0ebd3345549fb5bf67118fce6550b8)

HOSTNAME=buildlet-linux-stretch-morecpu-rn83cdb7c, so GOMAXPROCS=16.

```
    pprof_test.go:541: runtime/pprof.cpuHog1: 502
    pprof_test.go:541: runtime/pprof.cpuHog3: 5050
    --- FAIL: TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof (0.00s)
        pprof_test.go:179: compare 1m19.992473s vs 50.5s
        pprof_test.go:191: CPU usage reports are too different (limit -10.0%, got -36.9%)
```

The single-goroutine portion saw 5.02 seconds of samples (ran a little long!), and the parallel portion saw 50.50 seconds of samples (also ran a little long, but on fewer than GOMAXPROCS threads?). This could look like "the cliff at 2.5 that we got with setitimer is now present at 10 vCPU, sometimes" ... at least, 50.50 seconds is only slightly above 10 x 5.00 seconds.

There were no other failures in this run.

---

[2021-09-27T18:58:29-8cfd8c3/linux-amd64-longtest](https://build.golang.org/log/f1b9a22b2fef09e5219f117ee3db86154064e99e)

HOSTNAME=buildlet-linux-stretch-morecpu-rn6ff4f0f, so GOMAXPROCS=16.

```
    pprof_test.go:541: runtime/pprof.cpuHog1: 500
    pprof_test.go:541: runtime/pprof.cpuHog3: 5083
    --- FAIL: TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof (0.00s)
        pprof_test.go:179: compare 1m19.99259s vs 50.83s
        pprof_test.go:191: CPU usage reports are too different (limit -10.0%, got -36.5%)
```

The single-goroutine portion saw 5.00 seconds of samples, and the parallel portion saw 50.83 seconds of samples (ran a little long, but on fewer than GOMAXPROCS threads?). This could look like "the cliff at 2.5 that we got with setitimer is now present at 10 vCPU, sometimes" ... at least, 50.83 seconds is only slightly above 10 x 5.00 seconds.

There were no other failures in this run.

**Comment by rhysh** [comment]
These failures look like three problems:

1. The failures on linux/amd64 for 8cfd8c3 and 5b90958 (both on 2021-09-27) are because I put the test too early in the CL chain. We discussed at https://github.com/golang/go/issues/35057#issuecomment-928212995. The first WSL failures are on those two commits as well.
2. Failures in `TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof`. These appear on `linux-riscv64`, `linux-amd64-wsl`, and `linux-ppc64`. The first of these was `2021-09-30T16:43:32-9112d29`, and they've occured regularly since.
3. Failures in `TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof`, coupled with failures in other CPU-profiling-related test cases. These appear on `linux-amd64-longtest` and `linux-386-longtest. The first of these was `2021-10-18T19:46:27-394a1ad`

---

Problem 1 is a couple bad commits while the CL stack was landing. I'll filter those out of the results.

```
$ grep -r -l 'FAIL\truntime/pprof' . | egrep '2021-(09|10)' | grep '/linux-' | egrep '(8cfd8c3|5b90958)' | xargs grep -H 'FAIL:' | grep -v TestBlockProfile | sed -n -e 's,:[ ]*--- FAIL:,,p' | awk '{print $2,$1}' | sort | column -t
TestCPUProfileMultithreadMagnitude                                 ./2021-09-27T18:58:29-8cfd8c3/linux-amd64-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-09-27T18:58:29-8cfd8c3/linux-amd64-wsl
TestCPUProfileMultithreadMagnitude                                 ./2021-09-27T18:58:36-5b90958/linux-amd64-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-09-27T18:58:36-5b90958/linux-amd64-wsl
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-09-27T18:58:29-8cfd8c3/linux-amd64-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-09-27T18:58:29-8cfd8c3/linux-amd64-wsl
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-09-27T18:58:36-5b90958/linux-amd64-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-09-27T18:58:36-5b90958/linux-amd64-wsl
```

---

Problem 2 often shows garbage-collection-related functions in the profile. If the GC is running during the test, it would increase the CPU time for the whole process as reported by getrusage(2) without increasing the amount of time in the cpuHog functions that the test examines. I plan to update the test to disable the GC while it runs.

```
$ grep -r -l 'FAIL\truntime/pprof' . | egrep '2021-(09|10)' | grep '/linux-' | egrep -v '(8cfd8c3|5b90958)' | xargs grep -H 'FAIL:' | grep -v TestBlockProfile | sed -n -e 's,:[ ]*--- FAIL:,,p' | awk '{print $2,$1}' | sort | column -t | grep -v longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-09-30T16:43:32-9112d29/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude                                 ./2021-10-01T04:30:22-822f349/linux-amd64-wsl
TestCPUProfileMultithreadMagnitude                                 ./2021-10-01T17:27:24-952df98/linux-riscv64-jsing
TestCPUProfileMultithreadMagnitude                                 ./2021-10-01T17:27:29-3357624/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude                                 ./2021-10-02T13:52:59-07a7c6f/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude                                 ./2021-10-05T16:06:49-060cd73/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude                                 ./2021-10-06T16:11:20-aecf4b1/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude                                 ./2021-10-08T14:08:12-5b9206f/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude                                 ./2021-10-08T17:32:31-d7ba1d2/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude                                 ./2021-10-11T12:53:11-34f7b1f/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude                                 ./2021-10-12T20:20:41-d032b2b/linux-amd64-wsl
TestCPUProfileMultithreadMagnitude                                 ./2021-10-15T20:29:48-85cbdda/linux-riscv64-jsing
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T18:56:08-07e5527/linux-riscv64-unleashed
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T21:05:11-1b24c9e/linux-riscv64-unleashed
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T22:47:48-2be5b84/linux-riscv64-unleashed
TestCPUProfileMultithreadMagnitude                                 ./2021-10-20T05:28:45-b091189/linux-riscv64-unleashed
TestCPUProfileMultithreadMagnitude                                 ./2021-10-20T20:58:15-73652af/linux-ppc64-buildlet
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-09-30T16:43:32-9112d29/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-01T04:30:22-822f349/linux-amd64-wsl
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-01T17:27:24-952df98/linux-riscv64-jsing
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-01T17:27:29-3357624/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-02T13:52:59-07a7c6f/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-05T16:06:49-060cd73/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-06T16:11:20-aecf4b1/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-08T14:08:12-5b9206f/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-08T17:32:31-d7ba1d2/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-11T12:53:11-34f7b1f/linux-riscv64-unmatched
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-12T20:20:41-d032b2b/linux-amd64-wsl
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-15T20:29:48-85cbdda/linux-riscv64-jsing
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-19T18:56:08-07e5527/linux-riscv64-unleashed
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-19T21:05:11-1b24c9e/linux-riscv64-unleashed
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-19T22:47:48-2be5b84/linux-riscv64-unleashed
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-20T05:28:45-b091189/linux-riscv64-unleashed
TestCPUProfileMultithreadMagnitude/serial_execution_OS_vs_pprof    ./2021-10-20T20:58:15-73652af/linux-ppc64-buildlet
```

---

Problem 3 is new in the last few days. When it shows up, it affects the single-goroutine CPU profiling tests too. It's only appeared on the longtest builders. Those have 16 hyperthreads / GOMAXPROCS=16.

The failures in single-goroutine tests take the form of zero or near-zero samples in the desired functions, and the failures in `TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof` take the form of the CPU profile reporting only 6 to 14 threads worth of CPU time. (The tests expects 16 threads worth of reporting, and its 10% error margin means that reports from 15 threads worth of work would be considered a success.)

It's interesting that the amount of CPU time reported in the profile is always very close to an integer number of threads (each running for 5 seconds). It's also interesting that the smallest report is 6 out of the 16 threads. I disabled the new timer_create-based profiler and ran the test on the linux-386-longtest builder, and found that setitimer is able to deliver enough SIGPROF events to the process to cover the work of 10 threads (https://storage.googleapis.com/go-build-log/d3ff14b2/linux-386-longtest_a3d8929c.log). My working theory on this failure mode is that on some fraction of the process's threads, the kernel delivers the signals from setitimer first, so the deliveries from timer_create are crowded out up to setitimer's maximum signal delivery rate.

I also find it suspicious that this failure mode is new in the last few days, but none of the nearby commits look related and I also haven't been able to reproduce the problem myself.

```
$ grep -r -l 'FAIL\truntime/pprof' . | egrep '2021-(09|10)' | grep '/linux-' | egrep -v '(8cfd8c3|5b90958)' | xargs grep -H 'FAIL:' | grep -v TestBlockProfile | sed -n -e 's,:[ ]*--- FAIL:,,p' | awk '{print $2,$1}' | sort | column -t | grep longtest
TestCPUProfileInlining                                             ./2021-10-19T07:45:46-ee92daa/linux-386-longtest
TestCPUProfileInlining                                             ./2021-10-19T11:52:34-bde0463/linux-386-longtest
TestCPUProfileInlining                                             ./2021-10-19T18:42:57-7999fd4/linux-386-longtest
TestCPUProfileInlining                                             ./2021-10-19T22:47:48-2be5b84/linux-386-longtest
TestCPUProfileInlining                                             ./2021-10-20T07:49:28-d2ec73c/linux-amd64-longtest
TestCPUProfileInlining                                             ./2021-10-20T19:29:12-4320949/linux-386-longtest
TestCPUProfileInlining                                             ./2021-10-20T20:58:15-73652af/linux-386-longtest
TestCPUProfileLabel                                                ./2021-10-18T19:46:27-394a1ad/linux-amd64-longtest
TestCPUProfileLabel                                                ./2021-10-18T21:21:54-6c0daa7/linux-386-longtest
TestCPUProfileLabel                                                ./2021-10-19T07:45:46-ee92daa/linux-386-longtest
TestCPUProfileLabel                                                ./2021-10-19T11:52:34-bde0463/linux-386-longtest
TestCPUProfileLabel                                                ./2021-10-19T22:47:48-2be5b84/linux-386-longtest
TestCPUProfileLabel                                                ./2021-10-20T00:05:22-4cc6a91/linux-386-longtest
TestCPUProfileLabel                                                ./2021-10-20T20:39:19-d419a80/linux-386-longtest
TestCPUProfileLabel                                                ./2021-10-20T20:58:15-73652af/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-18T19:46:27-394a1ad/linux-amd64-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-18T21:21:54-6c0daa7/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T07:45:46-ee92daa/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T11:52:34-bde0463/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T16:19:27-f92a358/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T17:01:35-254c497/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T18:42:57-7999fd4/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-19T22:47:48-2be5b84/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-20T00:05:22-4cc6a91/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-20T07:49:28-d2ec73c/linux-amd64-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-20T19:29:12-4320949/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-20T20:39:19-d419a80/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-20T20:39:25-e904928/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-20T20:58:15-73652af/linux-386-longtest
TestCPUProfileMultithreadMagnitude                                 ./2021-10-21T15:42:41-8b0bea9/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-18T19:46:27-394a1ad/linux-amd64-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-18T21:21:54-6c0daa7/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-19T07:45:46-ee92daa/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-19T11:52:34-bde0463/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-19T16:19:27-f92a358/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-19T17:01:35-254c497/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-19T18:42:57-7999fd4/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-19T22:47:48-2be5b84/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-20T00:05:22-4cc6a91/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-20T07:49:28-d2ec73c/linux-amd64-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-20T19:29:12-4320949/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-20T20:39:19-d419a80/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-20T20:39:25-e904928/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-20T20:58:15-73652af/linux-386-longtest
TestCPUProfileMultithreadMagnitude/parallel_execution_OS_vs_pprof  ./2021-10-21T15:42:41-8b0bea9/linux-386-longtest
TestCPUProfileRecursion                                            ./2021-10-18T21:21:54-6c0daa7/linux-386-longtest
TestCPUProfileRecursion                                            ./2021-10-19T07:45:46-ee92daa/linux-386-longtest
TestCPUProfileRecursion                                            ./2021-10-19T11:52:34-bde0463/linux-386-longtest
TestCPUProfileRecursion                                            ./2021-10-19T22:47:48-2be5b84/linux-386-longtest
TestCPUProfileRecursion                                            ./2021-10-20T00:05:22-4cc6a91/linux-386-longtest
TestCPUProfileRecursion                                            ./2021-10-20T20:58:15-73652af/linux-386-longtest
TestMathBigDivide                                                  ./2021-10-18T21:21:54-6c0daa7/linux-386-longtest
TestMathBigDivide                                                  ./2021-10-19T07:45:46-ee92daa/linux-386-longtest
TestMathBigDivide                                                  ./2021-10-19T16:19:27-f92a358/linux-386-longtest
TestMathBigDivide                                                  ./2021-10-20T20:58:15-73652af/linux-386-longtest
TestMorestack                                                      ./2021-10-18T19:46:27-394a1ad/linux-amd64-longtest
TestMorestack                                                      ./2021-10-18T21:21:54-6c0daa7/linux-386-longtest
TestMorestack                                                      ./2021-10-19T07:45:46-ee92daa/linux-386-longtest
TestMorestack                                                      ./2021-10-19T11:52:34-bde0463/linux-386-longtest
TestMorestack                                                      ./2021-10-20T20:39:19-d419a80/linux-386-longtest
TestTimeVDSO                                                       ./2021-10-18T19:46:27-394a1ad/linux-amd64-longtest
TestTimeVDSO                                                       ./2021-10-18T21:21:54-6c0daa7/linux-386-longtest
TestTimeVDSO                                                       ./2021-10-19T07:45:46-ee92daa/linux-386-longtest
TestTimeVDSO                                                       ./2021-10-19T11:52:34-bde0463/linux-386-longtest
TestTimeVDSO                                                       ./2021-10-19T16:19:27-f92a358/linux-386-longtest
TestTimeVDSO                                                       ./2021-10-20T00:05:22-4cc6a91/linux-386-longtest
TestTimeVDSO                                                       ./2021-10-21T15:42:41-8b0bea9/linux-386-longtest
```

```
$ find . -name linux-386-longtest | egrep -v '(8cfd8c3|5b90958)' | egrep '2021-(09|10)' | sort | xargs egrep 'pprof_test.go.*compare'
./2021-10-18T21:21:54-6c0daa7/linux-386-longtest:        pprof_test.go:179: compare 1m19.983449s vs 29.94s
./2021-10-19T07:45:46-ee92daa/linux-386-longtest:        pprof_test.go:179: compare 1m19.558967s vs 34.71s
./2021-10-19T11:52:34-bde0463/linux-386-longtest:        pprof_test.go:179: compare 1m19.932465s vs 34.89s
./2021-10-19T16:19:27-f92a358/linux-386-longtest:        pprof_test.go:179: compare 1m19.987855s vs 59.95s
./2021-10-19T17:01:35-254c497/linux-386-longtest:        pprof_test.go:179: compare 1m19.980093s vs 1m4.9s
./2021-10-19T18:42:57-7999fd4/linux-386-longtest:        pprof_test.go:179: compare 1m19.950427s vs 1m4.86s
./2021-10-19T22:47:48-2be5b84/linux-386-longtest:        pprof_test.go:179: compare 1m19.920228s vs 1m4.86s
./2021-10-20T00:05:22-4cc6a91/linux-386-longtest:        pprof_test.go:179: compare 1m19.976284s vs 54.89s
./2021-10-20T19:29:12-4320949/linux-386-longtest:        pprof_test.go:179: compare 1m19.86348s vs 59.83s
./2021-10-20T20:39:19-d419a80/linux-386-longtest:        pprof_test.go:179: compare 1m19.975459s vs 1m9.87s
./2021-10-20T20:39:25-e904928/linux-386-longtest:        pprof_test.go:179: compare 1m18.992738s vs 44.33s
./2021-10-20T20:58:15-73652af/linux-386-longtest:        pprof_test.go:179: compare 1m19.982008s vs 1m4.89s
./2021-10-21T15:42:41-8b0bea9/linux-386-longtest:        pprof_test.go:179: compare 1m19.972142s vs 1m4.91s
```

**Comment by rhysh** [comment]
Thanks @prattmic . I've got 1171 runs and 0 failures now running https://github.com/torvalds/linux/commit/b6b178e38f40f34842b719a8786d346d4cfec5dc plus a patch to clear `tsk->posix_cputimers_work.scheduled` in `copy_process`.

What is the right way to move forward on fixing the kernel bug? I don't lurk on LKML, so don't know the norms to follow or to expect there. I'm open to learning, but handing off to someone with more facility there would be fine too.

---

It looks like `tsk->posix_cputimers_work.work` gets initialized only once, via `posix_cputimers_init_work` called by `init_timers` called by `start_kernel`. (My test run of b6b178e plus a patch to `posix_cputimers_init_work`, to set `scheduled` at the same time as `work`, continues to see the ~1% failure rate.) So I don't think it's correct to clear the `work` field at any time. It looks like it's only a function pointer, and that the job itself is an entry in `tsk->task_works` which as you said gets cleared during `copy_process`.

More questions I'd have re an LKML conversation are "which function should do the clearing", "or if it's copy_process, where", "or is there a way to keep all the code close to posix-cpu-timers.c, such as recording the thread id where the job is enqueued instead of true/false, so no clearing is required", and "does it need a reproducer written in C".

---

There's also the question of what this means for Go. Should we disable the tests unconditionally on linux/386 and linux/amd64 (because the kernel feature that introduced the bug is implemented for x86), or is there a good mechanism and precedent for disabling tests on particular kernel versions? Should we disable profiling when we clone threads? Is this worth an update to #13841?

Or: do we expect most distros will pick up the kernel bugfix in the next three months, so this will all shake out before the Go 1.18 release.

I also haven't addressed the flakiness on RISC-V and WSL. Those look related to CPU time spent on GC (which lead to an imbalance between rusage and the expected profile samples for the application code), so I plan to change the test to disable GC.

**Comment by prattmic** [comment]
> What is the right way to move forward on fixing the kernel bug? I don't lurk on LKML, so don't know the norms to follow or to expect there. I'm open to learning, but handing off to someone with more facility there would be fine too.

No problem, I will engage with them and CC you on the thread.

> Or: do we expect most distros will pick up the kernel bugfix in the next three months, so this will all shake out before the Go 1.18 release.

As you note earlier, this problem affects setitimer as well, so it isn't just a 1.18 problem. If we add a workaround, then it should likely be backported as well. I need to think a bit more about whether this is severe enough for a workaround like disabling profiling during clone.

**Comment by rhysh** [comment]
> No problem, I will engage with them and CC you on the thread.

Thanks @prattmic . And yes, the problem continues to reproduce in yesterday's release of 5.15.

**Comment by prattmic** [comment]
I've sent a patch for this upstream: https://lore.kernel.org/lkml/20211101210615.716522-1-mpratt@google.com/T/

**Comment by prattmic** [comment]
That patch has been merged into the -tip tree, timers/urgent branch, so it should be making its way in.

https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?id=ca7752caeaa70bd31d1714af566c9809688544af

**Comment by bcmills** [comment]
so... do we think this patch will hit the Go builders soon, or do we need some kind of workaround in the interim?

**Comment by prattmic** [comment]
From discussion today, I'm planning to write a general workaround for tip (but not backport), as the user impact seems fairly minimal.

In addition, we can mark the tests flaky until the workaround is available.

**Comment by rhysh** [comment]
That's great news, nice work on the patch and explanation to LKML. The commit in timers/urgent is working well for me so far:
```
Linux ip-172-31-61-152 5.15.0-custom-ca7752c #1 SMP PREEMPT Tue Nov 2 16:33:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
7h50m10s: 2778 runs so far, 0 failures
```

@prattmic , it sounds like you're on top of the whole picture of followup related to the kernel bug. Thank you! Are there parts of it that you're waiting on from me?

For the other half of the flakiness (the serial portion, on RISC-V and WSL), I sent https://golang.org/cl/359934 . I'll ping on that CL for a second reviewer.

**Comment by prattmic** [comment]
> Are there parts of it that you're waiting on from me?

No, though I'm working on a workaround to disable the timers during clone and would like your review of that once it is ready.

**Comment by bcmills** [comment]
Does the kernel bug also explain the recent flakes in `TestCgoPprof`?

`greplogs --dashboard -md -l -e '(?ms)FAIL: TestCgoPprof.*cpuHog traceback missing' --since=2020-06-16`

<details>

[2021-11-04T23:56:29-0e5f287/linux-amd64-buster](https://build.golang.org/log/b9f0dfaf3795fc80614582f7cdacb9cd107e0352)
[2021-11-02T21:18:44-80065cf/linux-amd64-staticlockranking](https://build.golang.org/log/122efc06496140ddb68445a56073e0bf965343dc)
[2021-11-02T16:12:23-599de4b/linux-amd64-sid](https://build.golang.org/log/a040b1bdb87f2b144551aae990f8f923e163dde8)
[2021-10-14T21:08:35-440b63b/linux-amd64-staticlockranking](https://build.golang.org/log/42bb604cde4cbeb798998bd46c37f776e9a2c2e7)
[2021-10-06T19:45:35-e38ec96/linux-ppc64le-buildlet](https://build.golang.org/log/e73acd94b9c1d36fe1ee6d15105cd1fab46442a3)
[2021-10-04T17:17:11-cc5e3de/linux-amd64-longtest](https://build.golang.org/log/c49b09355a866cb23c598a4253811a24ceb150fd)
[2021-09-28T20:54:26-b8a6017/linux-amd64-jessie](https://build.golang.org/log/2bf2205e7f4a75eb6d9abe506331ff693e8292bd)

</details>

(The symptoms of these failures match #37201, but that appeared to have been fixed in May 2020.)

**Comment by rhysh** [comment]
@bcmills this particular kernel bug wouldn't explain the failure on `linux-amd64-longtest` on October 4 (well before October 18).

Yes, those failures look like what's described in #37201. That issue closed when a particular source of signals was eliminated; there may be another (less frequent) source of signals.

The October 4 failure shows `Duration: 200.88ms, Total samples = 10ms ( 4.98%)`. Stopping the CPU profiler seems to involve 200ms worth of sleeps, so I take that to mean that it got stopped almost immediately, and due to the "traceback function called 2 times"  condition rather than "test ran for one second" (https://github.com/golang/go/blob/cc5e3de593afca73cf1b4d732ddceffb2837b390/src/runtime/testdata/testprogcgo/pprof.go#L89). Samples covering 10ms, at 100Hz, means 1 sample. That's consistent with the process receiving signals in addition to SIGPROF, causing the test to exit early.

**Comment by rhysh** [comment]
Turning the GC off during the test didn't resolve all sources of flakes on riscv64: https://build.golang.org/log/088b21dcceb4774c2f5ed27a7929611730fc0192 is a recent failure that shows 39 samples in `0x77188 (runtime.sigtramp:429)`.

That seems like a strange function to show up in CPU profiles (https://github.com/golang/go/blob/6ba68a0581d537edfd6eb2e5675c408f18eb4eed/src/runtime/sys_linux_riscv64.s#L429). How would it be able to get a signal while _already handling_ a signal?

I also see that it's a single PC. Maybe it's in the profile via sigprofNonGoPC (https://github.com/golang/go/blob/6ba68a0581d537edfd6eb2e5675c408f18eb4eed/src/runtime/signal_unix.go#L429), which would mean `g==nil`. FWIW, that also means (from validSIGPROF) that it's from setitimer rather than timer_create.

If that's the case, what would get in the way of finding `g` in signal handlers on riscv64? The PC isn't in the VDSO: we see the result of `c.sigpc()` (`0x77188`), and if `inVDSOPage` works correctly that should return false when `sigFetchG` calls it. So the `g` should come from `getg()`. It looks like that's X27 (https://github.com/golang/go/blob/go1.17.3/src/cmd/compile/internal/ssa/gen/RISCV64Ops.go#L27), which is supposed to get preserved across calls / not clobbered (if I've found the right doc and am reading it correctly: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc#integer-register-convention).

I'm not sure where to go from here. Are there RISCV or ABI experts who might have a theory, or is there a way I can access hardware to debug?

**Comment by prattmic** [comment]
> > Are there parts of it that you're waiting on from me?
> 
> No, though I'm working on a workaround to disable the timers during clone and would like your review of that once it is ready.

I tried this out, but unfortunately, it does not actually work. Just looking at clearing ITIMER_PROF, the behavior is:

* setitimer sets [tsk->signal->posix_cputimers.bases[CPUCLOCK_PROF].nextevt = 0](https://elixir.bootlin.com/linux/v5.15.1/source/kernel/time/posix-cpu-timers.c#L1447) and [tsk->signal->it[CPUCLOCK_PROF]](https://elixir.bootlin.com/linux/v5.15.1/source/kernel/time/itimer.c#L193).
* In the timer interrupt, [run_posix_cpu_timers](https://elixir.bootlin.com/linux/v5.15.1/source/kernel/time/posix-cpu-timers.c#L1384) will call `__run_posix_cpu_timers`, setting `tsk->posix_cputimers_work.scheduled = true` [as long there is a single tick of user/system time](https://elixir.bootlin.com/linux/v5.15.1/source/kernel/time/posix-cpu-timers.c#L1138).
  * Note here that `expiry_cache_is_inactive` looks for `nextevt == U64_MAX`, not 0.

This is normally fine, as `handle_posix_cpu_timers` won't actually send a signal since `tsk->signal->it[CPUCLOCK_PROF]` is cleared, but simply scheduling the work is sufficient for this bug.

As an alternative to disabling the timers around clone, one wonders if perhaps we can set them to an extremely high value so that `fastpath_timer_check` will determine that none have expired? No, unfortunately not, as setting the timer only changes `nextexp` if the new value is less than the old value (to avoid walking the entire linked list to search for the next expiration time). Instead, the timer work is scheduled whenever the next expiration would occur, at which point the list is walked and next expiration updated.

Thus, I don't currently have a workaround idea beyond setting the timers and spinning until the next expiration before clone (which is obviously unacceptable). I'm all ears to other ideas, otherwise I'll just disable this test for now.

**Comment by rhysh** [comment]
I see some options to hit the bug less often, but not to work around it entirely.

First of all, the initial timer expiration in the new profiler is chosen from a uniform random distribution in the range [`[1, 10_000_000] ns`](https://github.com/golang/go/blob/84277bfd07dad771b9978149bdaed8aa16ed8982/src/runtime/os_linux.go#L640). In `TestCPUProfileMultithreadMagnitude`, many goroutines are ready to work right away, so we expect the runtime to create new threads, for each new thread to create a profiling timer for itself, and then to immediately create an additional new thread (until there's an M for each P). In that scenario, pushing back the earliest possible timer_create expiration (e.g. uniformly distributed over 10–90% of the 10ms target instead of 0–100%, keeping the mean the same) could make the expiration _less likely_ to coincide with a call to clone for that type of workload.

Another option, we could use the template thread every time the runtime needs a new M (and not enable per-thread profiling for that thread). That thread would still be at risk of doing the buggy bookkeeping for the process-wide setitimer. And there's probably a good performance reason to not do this all the time already.

**Comment by prattmic** [comment]
re @rhysh 

Similar to your option (1), we do have a few tests that use `runtime_test.warmupScheduler` to get the thread count up before starting tests. We could do that for `TestCPUProfileMultithreadMagnitude`, but I don't think it is worth it vs just waiting.

Re: option (2), again since that is only a partial mitigation, I'd rather avoid making such a change unless this becomes really problematic.

**Comment by prattmic** [comment]
With http://golang.org/cl/363214 submitted, I believe we have addressed the flakiness on builders due to the Linux issue.

Additionally, all failures not related to the Linux kernel bug have been split into other issues (please correct me if I forgot some, in which case they should be moved).

Thus, this issue is done for now, but I'll leave open to track the upstream Linux fix rollout and ultimately dropping/adjusting the kernel version checks.

**Comment by prattmic** [comment]
Patch update: my upstream patch has been [merged](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=622c72b651c85cb55bae147debc1a2fae0189b53) into Linus' tree and queued for backport to [5.10](https://git.kernel.org/pub/scm/linux/kernel/git/stable/stable-queue.git/tree/queue-5.10/posix-cpu-timers-clear-task-posix_cputimers_work-in-copy_process.patch) and [5.14](https://git.kernel.org/pub/scm/linux/kernel/git/stable/stable-queue.git/tree/queue-5.14/posix-cpu-timers-clear-task-posix_cputimers_work-in-copy_process.patch).

**Comment by prattmic** [comment]
This fix is now released in:

* [5.10.80](https://lore.kernel.org/all/163724420148165@kroah.com/)
* [5.14.19](https://lore.kernel.org/all/163714476814814@kroah.com/)
* [5.15.3](https://lore.kernel.org/all/163726325011777@kroah.com/)

For reference, I took a look at some distros that I expect off the top of my head are widely used. Reminder this bug affects Linux 5.9 through 5.15 inclusive.

* RHEL 8: Linux 4.18, unaffected.
* Debian oldstable (buster): Linux 4.19, unaffected
* Debian stable (bullseye): Linux 5.10.70 currently, appears to update patch release every ~3 months.
* Debian testing (bookworm): Linux 5.14.16 currently, appears to update patch release monthly.
* Ubuntu 20.04 LTS: Linux 5.4, unaffected
* Ubuntu 22.10: Linux 5.13. This is no longer supported upstream, so it was not patched. It is unclear what their backport policy is, but the [release notes](https://launchpad.net/ubuntu/impish/+source/linux) do list all the changes they integrate.
* Amazon Linux 2: Linux 4.14, unaffected.
* Google Container-Optimized OS M93 LTS: Linux 5.10.68 currently. Updates patch release in "LTS Refresh" releases every 3 months.
* Google Container-Optimized OS M89 LTS: Linux 5.4, unaffected.

Overall this looks pretty promising to me that most of the ecosystem should get the patches fairly quickly.

**Comment by bcmills** [comment]
Do we also need to apply the test-skip to some of the other `runtime` and/or `runtime/pprof` tests mentioned in https://github.com/golang/go/issues/49065#issuecomment-946825081?

For example, I still see some recent `TestMorestack` failures, and there was a large uptick in those starting Oct. 18 as well:

`greplogs --dashboard -md -l -e 'FAIL: TestMorestack' --since=2021-07-01`

<details>

[2021-11-18T02:16:39-353cb71/linux-386-longtest](https://build.golang.org/log/27320f7a31d057895f465749e023196729f344f2)
[2021-11-11T20:25:07-eb68e33/linux-amd64-longtest](https://build.golang.org/log/a61ddaa0fa6c01e8b6e42161f0272ca958152724)
[2021-11-10T20:45:04-a881409/linux-386-longtest](https://build.golang.org/log/1a7489b433b3c4fb7fa82174b067321bfa30a937)
[2021-11-10T18:06:32-097aaa9/linux-386-longtest](https://build.golang.org/log/ee21b652230255f3ab68e40366d635b64c83e17a)
[2021-11-09T22:11:33-4aa0746/linux-386-longtest](https://build.golang.org/log/d8652f33732091011cea398f93b545b8d9462f52)
[2021-11-09T20:10:44-74b9939/linux-386-longtest](https://build.golang.org/log/d751f0b4332604283ee854370ffddea7c1d63111)
[2021-11-09T18:03:59-36dbf7f/linux-386-longtest](https://build.golang.org/log/49303f21ca9aca0b930381e3aed6d404d727305b)
[2021-11-08T21:45:43-cc49178/linux-amd64-longtest](https://build.golang.org/log/32ea3df61189af23d688cf47e10181a6ee1a56f7)
[2021-11-08T18:54:21-830b393/linux-386-longtest](https://build.golang.org/log/85b24e1bb2a610fa8c0ea4f747192dd647440bf1)
[2021-11-07T04:57:22-9e6ad46/linux-amd64-longtest](https://build.golang.org/log/f523a728c6c92bce14712815f060e063b143d63a)
[2021-11-06T19:41:10-565f457/linux-amd64-longtest](https://build.golang.org/log/f6b6ed25d4348f0ddc75fef3a1db2696cdbbd4a8)
[2021-11-05T22:56:55-2f71c86/linux-amd64-longtest](https://build.golang.org/log/fbd8fdc62a34e7605d97232c513d692c3f39af7f)
[2021-11-05T22:30:17-b07c41d/linux-386-longtest](https://build.golang.org/log/eb9927bcb8358bb3f8707ec04d6f2705d23f9e89)
[2021-11-05T19:14:22-fa16efb/linux-386-longtest](https://build.golang.org/log/acaced40841431ddda21daa73d77aa6bfe045f3e)
[2021-11-05T17:46:59-e48e4b4/linux-amd64-longtest](https://build.golang.org/log/88df6286352fc1cbfa4736ccc78ccd8858976578)
[2021-11-05T00:52:04-1c4cfd8/linux-amd64-longtest](https://build.golang.org/log/53f125f3056386345fa397ff0d5cba68551813a2)
[2021-11-04T21:40:51-2c32f29/linux-amd64-longtest](https://build.golang.org/log/af0b4a8bce675ce944285fab19b1ebc935776d04)
[2021-11-04T15:43:59-6ba68a0/linux-386-longtest](https://build.golang.org/log/759676a044a9a98b109b82f2b4f8a13028b1860d)
[2021-11-04T02:57:37-9e4e23c/linux-386-longtest](https://build.golang.org/log/9197f06ca8d7f6af06fe6963846545bf98cd7b97)
[2021-11-04T00:46:41-e72d715/linux-386-longtest](https://build.golang.org/log/66751c116b12641454793d2a3aa6ee52b183ceca)
[2021-11-04T00:15:44-91962df/linux-386-longtest](https://build.golang.org/log/af6ffdfdfcb6af4951c78e7a6b158d34c6193e80)
[2021-11-03T05:00:00-519c0a2/linux-386-longtest](https://build.golang.org/log/28d015f95c198f23f4e5a67ebf9ab20598f861a3)
[2021-11-02T18:35:29-433ba58/linux-386-longtest](https://build.golang.org/log/91c604199087a92a89b2c80c9b8818dabda5a066)
[2021-11-02T15:15:09-da7173a/linux-386-longtest](https://build.golang.org/log/df1f0396606b8a37eeec7a769e2a0b17e67d575d)
[2021-11-02T01:28:01-a59e332/linux-386-longtest](https://build.golang.org/log/0fb5ac7a0d936ed1add7579f35038dde7cb3210d)
[2021-11-01T16:31:02-2bcf1c0/linux-386-longtest](https://build.golang.org/log/0274ad1241c0105ff5276b1e9502e3ea14b0c3df)
[2021-11-01T16:30:49-80bedb8/linux-386-longtest](https://build.golang.org/log/64eac015d4f4feab8ba561e139d6f8732091b4ce)
[2021-10-31T02:24:29-0bef30d/linux-386-longtest](https://build.golang.org/log/b178e5d015b96848825c79a47f5806edc477720e)
[2021-10-30T18:30:05-f582778/linux-386-longtest](https://build.golang.org/log/69a37ec29bfd62f7efed42a8198e64e4fa55fa66)
[2021-10-29T23:15:00-8dfb447/linux-386-longtest](https://build.golang.org/log/37383640e3e22ae23addb22f0c5bac15a404de7b)
[2021-10-29T18:34:24-903f313/linux-386-longtest](https://build.golang.org/log/4cefd34def45e2feaba1cf9d65e8ef77afe39071)
[2021-10-26T22:05:53-80be4a4/linux-386-longtest](https://build.golang.org/log/e9f3529400d092b189544d6be9ae2fee64045478)
[2021-10-25T17:02:16-52b10ab/linux-386-longtest](https://build.golang.org/log/edab701481385becf378b9f6424780670e159996)
[2021-10-23T20:23:35-8dbf3e9/linux-386-longtest](https://build.golang.org/log/9bba4795fd0f2695dea014a0fa434f42b528f1c5)
[2021-10-23T16:18:19-16318df/linux-386-longtest](https://build.golang.org/log/b372b4043bd23e2a81d483ac653a70ca510874c3)
[2021-10-22T21:12:07-76dd01f/linux-386-longtest](https://build.golang.org/log/066a582d1024c822e5a60514a2f193f1d30f9448)
[2021-10-20T20:39:19-d419a80/linux-386-longtest](https://build.golang.org/log/96eb074e0c2b1316f2840102a08201e09014d820)
[2021-10-19T11:52:34-bde0463/linux-386-longtest](https://build.golang.org/log/d8cfe4fa827b32911bc32af3ecc28ae2fe7c15c7)
[2021-10-19T07:45:46-ee92daa/linux-386-longtest](https://build.golang.org/log/96de4e4aa99b8f02198122daf296bfcbbdb76fd7)
[2021-10-18T21:21:54-6c0daa7/linux-386-longtest](https://build.golang.org/log/cd243d02a622b97f575d8f938e3029c45419790a)
[2021-10-18T19:46:27-394a1ad/linux-amd64-longtest](https://build.golang.org/log/98427a8ba9869e3d2dea30e65757798c0eb00791)

</details>

**Comment by prattmic** [comment]
Yes, we probably should. `TestMorestack` for example probably doesn't create as many threads as `TestCPUProfileMultithreadMagnitude`, but it could still be affected if unlucky.

Most of the pprof tests are probably at least theoretically affected, but I think we only want to add skips to tests that are particularly affected, lest we lose all pprof coverage.

Perhaps better would be to pin the builders to COS M89, which is unaffected, until M93 is updated to include the patch.

**Comment by bcmills** [comment]
> Perhaps better would be to pin the builders to COS M89, which is unaffected, until M93 is updated to include the patch.

@golang/release, is that feasible to do in the short term?

**Comment by prattmic** [comment]
IIUC, that could be achieved by changing https://cs.opensource.google/go/x/build/+/master:buildlet/gce.go;l=461;drc=05d22632dd9ae9f87a48a12960617e223cecdf9c?ss=go from `cos-stable` to `cos-89-lts` (https://cloud.google.com/container-optimized-os/docs/release-notes#lts_image_families), though that would affect every single builder using COS, which may or may not be desirable.

**Comment by heschi** [comment]
Is there anything left to do on this now that the builders are updated? @dmitshur points out it might deserve a release note.

**Comment by prattmic** [comment]
There is no remaining immediate work to be done beyond tracking rollout of upstream fixes.

Perhaps adding a release note would be a good place to note this bug, though I want to emphasize that this is not a new issue with Go 1.18. It affects old releases as well.

**Comment by heschi** [comment]
Feel free to send a release note :)

Closing this then.

**Comment by prattmic** [comment]
As of 2022-01-18 all of the distros I tracked in https://github.com/golang/go/issues/49065#issuecomment-973203258 have released a fixed kernel.

As a summary, this bug was:
* Introduced in **Linux 5.9**.
* Fixed in **Linux 5.10.80**.
* Fixed in Ubuntu's **Linux 5.13.0-24.24**.
* Fixed in **Linux 5.14.19**.
* Fixed in **Linux 5.15.3**.
* Fixed in **Linux 5.16**+.
